pydal/objects.py

```diff
@@ -1039,6 +1053,152 @@ def drop_index(self, name):
+            self[field.name] = field
+        self.ALL = SQLALL(self)
+
+    @property
+    def fields(self):
+        return self._fields
+
+    def update(self, *args, **kwargs):
+        raise RuntimeError("update() method not supported")
+
+    def __getitem__(self, key):
+        try:
+            return getattr(self, key)
+        except AttributeError:
+            raise KeyError(key)
+
+    def __setitem__(self, key, value):
+        self.__dict__[str(key)] = value
+
+    def __call__(self):
+        adapter = self._db._adapter
+        colnames, sql = self._compile()
+        cache = self._attributes.get('cache', None)
+        if cache and self._attributes.get('cacheable', False):
+            return adapter._cached_select(cache, sql, self._fields,
+                self._attributes, colnames)
+        return adapter._select_aux(sql, self._qfields, self._attributes,
+            colnames)
+
+    def __setattr__(self, key, value):
+        if key[:1] != '_' and key in self:
+            raise SyntaxError(
+                'Object exists and cannot be redefined: %s' % key)
+        self[key] = value
+
+    def __iter__(self):
+        for fieldname in self.fields:
+            yield self[fieldname]
+
+    def __repr__(self):
+        return '<Select (%s)>' % ', '.join(map(str, self._qfields))
+
+    def __str__(self):
+        return self._compile(with_alias=(self._tablename is not None))[1]
+
+    def with_alias(self, alias):
+        other = copy.copy(self)
+        other['ALL'] = SQLALL(other)
+        other['_tablename'] = alias
+        for fieldname in other.fields:
+            other[fieldname] = copy.copy(other[fieldname])
+            other[fieldname]._tablename = alias
+            other[fieldname].tablename = alias
+            other[fieldname]._table = other
+            other[fieldname].table = other
+        return other
+
+    def on(self, query):
+        if not self._tablename:
+            raise SyntaxError("Subselect must be aliased for use in a JOIN")
+        return Expression(self._db, self._db._adapter.dialect.on, self, query)
+
+    def _compile(self, outer_scoped=[], with_alias=False):
+        if not self._correlated:
+            outer_scoped = []
+        if outer_scoped or not self._sql_cache:
+            adapter = self._db._adapter
+            attributes = self._attributes.copy()
+            attributes['outer_scoped'] = outer_scoped
+            colnames, sql = adapter._select_wcols(self._query, self._qfields,
+                **attributes)
+            # Do not cache when the query may depend on external tables
+            if not outer_scoped:
+                self._colnames_cache, self._sql_cache = colnames, sql
+        else:
+            colnames, sql = self._colnames_cache, self._sql_cache
+        if with_alias and self._tablename is not None:
+            alias = self._db._adapter.dialect.quote(self._tablename)
+            if 'Oracle' in str(type(self._db._adapter)):
```

Comment:
@nextghost you should use dialect's `_as` method instead of checking the adapter type